modules/packages/DistributedBag.chpl

```diff
@@ -326,958 +357,809 @@ module DistributedBag {
-          delete tmp;
+            if splitreq then return (REMOVE_FAST_EXIT, default);
 
-          if headBlock == nil then tailBlock = nil;
-        }
-      }
+            phase = REMOVE_GLOBAL_STEAL;
+          }
 
-      return arr;
-    }
+          /*
+            Global steal: if the local steal failed, the calling task try to steal
+                          elements globally, i.e. from another bag instance. Similarly
+                          to the previous case, we choose the victim tasks by iterating
+                          over the locales, and then over their segments. When an
+                          eligible locale is found, we try to steal one element from
+                          each of its segments. Since the global WS is a heavy-weigth
+                          operation, the steal-1 strategy is no more appropriate, and
+                          we steal each segment in order to not create load-unbalance
+                          between them.
+          */
+          when REMOVE_GLOBAL_STEAL {
+            // fast exit for single-node execution
+            if (numLocales == 1) then return (REMOVE_FAIL, default);
 
-    inline proc ref takeElement() {
-      if isEmpty {
-        var default: eltType;
-        return (false, default);
-      }
+            // lock the global steal operation
+            if !globalStealInProgress.compareAndSwap(false, true) {
+              return (REMOVE_FAST_EXIT, default);
+            }
 
-      if headBlock!.isEmpty {
-        halt("DistributedBag Internal Error: Iterating over 1 element with headBlock empty but nElems is ", nElems.read());
-      }
+            const parentPid = parentHandle.pid;
+            var stolenElts: list(eltType);
 
-      var elem = headBlock!.pop();
-      nElems.sub(1);
+            // iterate over the victim locales
+            for victimLocaleId in victim(numLocales, here.id, "rand", 1) { //numLocales-1) {
+              on Locales[victimLocaleId] {
+                var targetBag = chpl_getPrivatizedCopy(parentHandle.type, parentPid).bag;
+                // iterate over the victim tasks
+                for victimTaskId in victim(here.maxTaskPar, -1, "rand", here.maxTaskPar) {
+                  ref targetSegment = targetBag!.segments[victimTaskId];
 
-      // Fix list if we consumed last one...
-      if headBlock!.isEmpty {
-        var tmp = headBlock;
-        headBlock = headBlock!.next;
-        delete tmp;
+                  targetSegment.globalSteal.write(true);
 
-        if headBlock == nil then tailBlock = nil;
-      }
+                  targetSegment.lock_block.readFE();
+                  // check eligibility
+                  if (distributedBagWorkStealingMinElts < targetSegment.nElts_shared.read()) {
+                    var (hasElt, elt): (bool, eltType) = targetSegment.stealElement();
 
-      return (true, elem);
-    }
+                    if hasElt {
+                      stolenElts.insert(0, elt);
```

Comment:
I suspect that such distributed list insertion is not particularly efficient. The `list` data type works better if you are appending; inserting at the `0`th position is `O(n)`. Worth at least a TODO comment.

Reply:
Added a TODO